glib-macros/src/clone.rs

```diff
@@ -1,952 +1,487 @@
-    fn from(f: &'a Group) -> Self {
-        let current_span = Some(f.span());
-        let mut inner = f.stream().into_iter().peekable();
-        let next_span = inner.peek().map(|n| n.span());
-        Self {
-            inner,
-            current_span,
-            next_span,
-        }
-    }
-}
-
-impl From<TokenStream> for PeekableProcIter {
-    fn from(f: TokenStream) -> Self {
-        let mut inner = f.into_iter().peekable();
-        let next_span = inner.peek().map(|n| n.span());
-        Self {
-            inner,
-            current_span: None,
-            next_span,
-        }
-    }
-}
-
-impl Iterator for PeekableProcIter {
-    type Item = TokenTree;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let next = self.inner.next()?;
-        self.current_span = Some(next.span());
-        self.next_span = self.inner.peek().map(|n| n.span());
-        Some(next)
-    }
-}
-
-impl PeekableProcIter {
-    fn peek(&mut self) -> Option<&TokenTree> {
-        self.inner.peek()
-    }
-
-    fn generate_error(&self, message: &str) -> TokenStream {
-        self.generate_error_with_span(message, self.current_span)
-    }
-
-    fn generate_error_with_next_span(&self, message: &str) -> TokenStream {
-        self.generate_error_with_span(message, self.next_span)
-    }
-
-    fn generate_error_with_span(&self, message: &str, span: Option<Span>) -> TokenStream {
-        let span = span.unwrap_or_else(Span::call_site);
-        // We generate a `compile_error` macro and assign it to the current span so the error
-        // displayed by rustc points to the right location.
-        let mut stream = TokenStream::new();
-        stream.extend(vec![TokenTree::Literal(Literal::string(message))]);
-
-        let mut tokens = vec![
-            TokenTree::Ident(Ident::new("compile_error", span)),
-            TokenTree::Punct(Punct::new('!', Spacing::Alone)),
-            TokenTree::Group(Group::new(Delimiter::Parenthesis, stream)),
-            TokenTree::Punct(Punct::new(';', Spacing::Alone)),
-        ];
-
-        for tok in &mut tokens {
-            tok.set_span(span);
-        }
-
-        let mut stream = TokenStream::new();
-        stream.extend(tokens);
-        let mut t = TokenTree::Group(Group::new(Delimiter::Brace, stream));
-        t.set_span(span);
-
-        let mut stream = TokenStream::new();
-        stream.extend(vec![t]);
-        stream
-    }
-}
-
-#[derive(Clone, Copy, Debug)]
-enum BorrowKind {
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub(crate) enum CaptureKind {
+    Watch,
     Weak,
     WeakAllowNone,
     Strong,
     ToOwned,
 }
 
-impl BorrowKind {
-    fn to_str(self) -> &'static str {
-        match self {
-            Self::Weak => "@weak",
-            Self::WeakAllowNone => "@weak-allow-none",
-            Self::Strong => "@strong",
-            Self::ToOwned => "@to-owned",
-        }
+impl CaptureKind {
+    pub(crate) fn kinds() -> &'static [&'static str] {
```

Comment:
Instead of checking like that, why not implement `TryFrom<str>` (or `TryFromStr`, never remember the name)? Like that you can make `Capture::new` take `CaptureKind` argument.